src/clients/http.rs

```diff
@@ -271,23 +291,191 @@ impl HttpClient {
+    fn make_span(&mut self, request: &Request<BoxBody<Bytes, hyper::Error>>) -> Span {
+        info_span!(
+            "client HTTP request",
+            request_method = request.method().to_string(),
+            request_path = request.uri().path().to_string(),
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClientOnRequest;
+
+impl OnRequest<BoxBody<Bytes, hyper::Error>> for ClientOnRequest {
+    fn on_request(&mut self, request: &Request<BoxBody<Bytes, hyper::Error>>, span: &Span) {
+        let _guard = span.enter();
+        info!(
+            trace_id = %trace::current_trace_id(),
+            method = %request.method(),
+            path = %request.uri().path(),
+            "started processing request",
+        );
+        info!(
+            monotonic_counter.incoming_request_count = 1,
+            request_method = request.method().as_str(),
+            request_path = request.uri().path()
+        );
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClientOnResponse;
+
+impl OnResponse<Incoming> for ClientOnResponse {
+    fn on_response(self, response: &hyper::Response<Incoming>, latency: Duration, span: &Span) {
+        let _guard = span.enter();
+        info!(
+            trace_id = %trace::current_trace_id(),
+            status = %response.status(),
+            duration_ms = %latency.as_millis(),
+            "finished processing request",
+        );
+
+        // On every response
+        info!(
+            monotonic_counter.client_response_count = 1,
+            response_status = response.status().as_u16(),
+            request_duration = latency.as_millis()
+        );
+        info!(
+            histogram.client_request_duration = latency.as_millis() as u64,
+            response_status = response.status().as_u16()
+        );
+
+        if response.status().is_server_error() {
+            // On every server error (HTTP 5xx) response
+            info!(
+                monotonic_counter.client_5xx_response_count = 1,
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else if response.status().is_client_error() {
+            // On every client error (HTTP 4xx) response
+            info!(
+                monotonic_counter.client_4xx_response_count = 1,
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else if response.status().is_success() {
+            // On every successful (HTTP 2xx) response
+            info!(
+                monotonic_counter.client_success_response_count = 1,
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else {
+            error!(
+                "unexpected HTTP client response status code: {}",
+                response.status().as_u16()
+            );
+        }
```

Comment:
I haven't tested the code, but this looks to be sending traces with duplicate data. 

The first trace, sent for every response, includes `response_status` and `request_duration` fields. The second trace, also sent for every response, also has a `response_status` field. I think the second trace can be dropped with `histogram.client_request_duration` moved up to the first trace, although I'm not familiar with the counters.

The if else blocks are also sending traces with `response_status` and `request_duration` fields rather than just incrementing their unique counters.

Reply:
I used the wrong terminology. Each `info!()` or other `event!()` variant produces a new "event", not a new "trace", within the context of a "span".